File: src/util/async/executor/ExecuteN.java
package util.async.executor;

import java.util.concurrent.Executor;

/**
 * Executor with an execution count limit.
 * <p/>
 * Guarantees the number of executions (irrelevant of the Runnable), as
 * one may wish for this executor to execute at most most n times.
 */
public class ExecuteN implements Executor {

	private long executed = 0;
	private final long max;

	/**
	 * @param limit maximum number of times this executor can execute any runnable
	 */
	public ExecuteN(long limit) {
		max = limit;
	}

	@Override
	public void execute(Runnable r) {
		if (executed<max) r.run();
		executed++;
	}

}

File: src/sp/it/pl/util/async/executor/ExecuteN.kt
package sp.it.pl.util.async.executor

import java.util.concurrent.Executor

/**
 * Executor with an execution count limit.
 *
 * Guarantees the number of executions (irrelevant of the [Runnable]), as
 * one may wish for this executor to execute at most n times.
 */
class ExecuteN : Executor {

    /** @param max maximum number of times this executor will [execute] */
    constructor(max: Long) { this.max = max }

    private val max: Long

    private var executed: Long = 0

    override fun execute(r: Runnable) {
        if (executed >= max)
            return
        r.run()
        executed++
    }

}